cluster/images/etcd/migrate/migrate.go

```diff
@@ -0,0 +1,208 @@
+		glog.Errorf("--target-storage is required")
+		os.Exit(1)
+	}
+	if opts.targetVersion == "" {
+		glog.Errorf("--target-version is required")
+		os.Exit(1)
+	}
+	if opts.dataDir == "" {
+		glog.Errorf("--data-dir is required")
+		os.Exit(1)
+	}
+	if opts.bundledVersionString == "" {
+		glog.Errorf("--bundled-versions is required")
+		os.Exit(1)
+	}
+
+	bundledVersions, err := ParseSupportedVersions(opts.bundledVersionString)
+	if err != nil {
+		glog.Errorf("Failed to parse --supported-versions: %v", err)
+	}
+	err = validateBundledVersions(bundledVersions, opts.binDir)
+	if err != nil {
+		glog.Errorf("Failed to validate that 'etcd-<version>' and 'etcdctl-<version>' binaries exist in --bin-dir '%s' for all --bundled-verions '%s': %v",
+			opts.binDir, opts.bundledVersionString, err)
+		os.Exit(1)
+	}
+
+	target := &EtcdVersionPair{
+		version:        MustParseEtcdVersion(opts.targetVersion),
+		storageVersion: MustParseEtcdStorageVersion(opts.targetStorage),
+	}
+
+	migrate(opts.name, opts.peerListenUrls, opts.peerAdvertiseUrls, opts.binDir, opts.dataDir, opts.etcdDataPrefix, opts.ttlKeysDirectory, opts.initialCluster, target, bundledVersions, opts.etcdServerArgs)
+}
+
+// migrate opens or initializes the etcd data directory, configures the migrator, and starts the migration.
+func migrate(name string, peerListenUrls string, peerAdvertiseUrls string, binPath string, dataDirPath string, etcdDataPrefix string, ttlKeysDirectory string,
+	initialCluster string, target *EtcdVersionPair, bundledVersions SupportedVersions, etcdServerArgs string) {
+
+	dataDir, err := OpenOrCreateDataDirectory(dataDirPath)
+	if err != nil {
+		glog.Errorf("Error opening or creating data directory %s: %v", dataDirPath, err)
+		os.Exit(1)
+	}
+
+	cfg := &EtcdMigrateCfg{
+		binPath:           binPath,
+		name:              name,
+		port:              mustFindAvailablePort(clientPortStart, clientPortEnd),
+		peerListenUrls:    peerListenUrls,
+		peerAdvertiseUrls: peerAdvertiseUrls,
+		etcdDataPrefix:    etcdDataPrefix,
+		ttlKeysDirectory:  ttlKeysDirectory,
+		initialCluster:    initialCluster,
+		supportedVersions: bundledVersions,
+		dataDirectory:     dataDirPath,
+		etcdServerArgs:    etcdServerArgs,
+	}
+	client, err := NewEtcdMigrateClient(cfg)
+	if err != nil {
+		glog.Errorf("Migration failed: %v", err)
+		os.Exit(1)
+	}
+	defer client.Close()
+
+	migrator := &Migrator{cfg, dataDir, client}
+
+	err = migrator.MigrateIfNeeded(target)
+	if err != nil {
+		glog.Errorf("Migration failed: %v", err)
+		os.Exit(1)
+	}
+}
+
+// mustFindAvailablePort searches the given port range for a port that the current process can listen on.
+func mustFindAvailablePort(rangeStart, rangeEnd uint64) uint64 {
+	var port uint64
+	for p := rangeStart; port == 0 && p <= rangeEnd; p++ {
+		ln, err := net.Listen("tcp", fmt.Sprintf(":%d", p))
+		if err == nil {
+			glog.Infof("port %d available, using it as temporary etcd port client port for migration", p)
+			ln.Close()
```

Comment:
Won't it cause problems?
Don't we need to wait a bit until port is available again after opening it?

Maybe if you close it properly not - just trying to ensure you thought about it.

Reply:
Good question. I did a bit of testing and reading and this looks safe. The `Close()` call does the right thing here and makes sure the TIME-WAIT don't prevent a subsequent listen fail.  Here's a single process example: https://play.golang.org/p/uc2LxCB6FV2. In the case here the etcd server process will listen after the port is found, so it's a slightly different situation, but my integration test are using this and all pass. https://github.com/phayes/freeport/blob/master/freeport.go also uses this approach.